dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/feedback/ObjectBundleValidationReport.java

```diff
@@ -27,70 +27,50 @@
  */
 package org.hisp.dhis.dxf2.metadata.objectbundle.feedback;
 
-import java.util.ArrayList;
 import java.util.HashMap;
-import java.util.List;
+import java.util.Iterator;
 import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
 
 import org.hisp.dhis.feedback.ErrorCode;
 import org.hisp.dhis.feedback.ErrorReport;
-import org.hisp.dhis.feedback.ObjectReport;
+import org.hisp.dhis.feedback.ErrorReportContainer;
 import org.hisp.dhis.feedback.TypeReport;
 
 import com.google.common.base.MoreObjects;
 
 /**
  * @author Morten Olav Hansen <mortenoh@gmail.com>
  */
-public class ObjectBundleValidationReport
+public class ObjectBundleValidationReport implements ErrorReportContainer, Iterable<TypeReport>
 {
-    private Map<Class<?>, TypeReport> typeReportMap = new HashMap<>();
-
-    public ObjectBundleValidationReport()
-    {
-    }
+    private final Map<Class<?>, TypeReport> typeReportMap = new HashMap<>();
 
     // -----------------------------------------------------------------------------------
     // Utility Methods
     // -----------------------------------------------------------------------------------
-
-    public List<ErrorReport> getErrorReportsByCode( Class<?> klass, ErrorCode errorCode )
+    public int getErrorReportsCountByCode( Class<?> klass, ErrorCode errorCode )
     {
-        List<ErrorReport> errorReports = new ArrayList<>();
-
-        if ( !typeReportMap.containsKey( klass ) )
-        {
-            return errorReports;
-        }
-
-        List<ObjectReport> objectReports = typeReportMap.get( klass ).getObjectReports();
-
-        for ( ObjectReport objectReport : objectReports )
-        {
-            List<ErrorReport> byCode = objectReport.getErrorReportsByCode().get( errorCode );
-
-            if ( byCode != null )
-            {
-                errorReports.addAll( byCode );
-            }
-        }
-
-        return errorReports;
+        TypeReport report = typeReportMap.get( klass );
+        return report == null ? 0 : report.getErrorReportsCount( errorCode );
     }
 
     public void addTypeReport( TypeReport report )
     {
         if ( report == null )
-            return;
-
-        if ( !typeReportMap.containsKey( report.getKlass() ) )
         {
-            typeReportMap.put( report.getKlass(), report );
             return;
         }
 
-        TypeReport typeReport = typeReportMap.get( report.getKlass() );
-        typeReport.merge( typeReport );
```

Comment:
Note the bug: `typeReport` is merged with itself instead of merging with `report`.